druid-handler/src/java/org/apache/hadoop/hive/druid/io/DruidRecordWriter.java

```diff
@@ -0,0 +1,266 @@
+      }
+    } else {
+      retVal = new SegmentIdentifier(
+              dataSchema.getDataSource(),
+              interval,
+              tuningConfig.getVersioningPolicy().getVersion(interval),
+              new LinearShardSpec(0)
+      );
+      pushSegments(Lists.newArrayList(currentOpenSegment));
+      currentOpenSegment = retVal;
+      return retVal;
+    }
+  }
+
+  private void pushSegments(List<SegmentIdentifier> segmentsToPush) {
+    try {
+      SegmentsAndMetadata segmentsAndMetadata = appenderator
+              .push(segmentsToPush, committerSupplier.get()).get();
+      final HashSet<String> pushedSegmentIdentifierHashSet = new HashSet<>();
+
+      for (DataSegment pushedSegment : segmentsAndMetadata.getSegments()) {
+        pushedSegmentIdentifierHashSet
+                .add(SegmentIdentifier.fromDataSegment(pushedSegment).getIdentifierAsString());
+        final Path segmentDescriptorOutputPath = DruidStorageHandlerUtils
+                .makeSegmentDescriptorOutputPath(pushedSegment, segmentsDescriptorDir);
+        DruidStorageHandlerUtils
+                .writeSegmentDescriptor(fileSystem, pushedSegment, segmentDescriptorOutputPath);
+
+        LOG.info(
+                String.format(
+                        "Pushed the segment [%s] and persisted the descriptor located at [%s]",
+                        pushedSegment,
+                        segmentDescriptorOutputPath
+                )
+        );
+      }
+
+      final HashSet<String> toPushSegmentsHashSet = new HashSet(
+              FluentIterable.from(segmentsToPush)
+                      .transform(new Function<SegmentIdentifier, String>() {
+                        @Nullable
+                        @Override
+                        public String apply(
+                                @Nullable SegmentIdentifier input
+                        ) {
+                          return input.getIdentifierAsString();
+                        }
+                      })
+                      .toList());
+
+      if (!pushedSegmentIdentifierHashSet.equals(toPushSegmentsHashSet)) {
+        throw new IllegalStateException(String.format(
+                "was asked to publish [%s] but was able to publish only [%s]",
+                Joiner.on(", ").join(toPushSegmentsHashSet),
+                Joiner.on(", ").join(pushedSegmentIdentifierHashSet)
+        ));
+      }
+
+      LOG.info(String.format("Published [%,d] segments.", segmentsToPush.size()));
+    } catch (InterruptedException e) {
+      LOG.error(String.format("got interrupted, failed to push  [%,d] segments.",
+              segmentsToPush.size()
+      ), e);
+      Thread.currentThread().interrupt();
+    } catch (IOException | ExecutionException e) {
+      LOG.error(String.format("Failed to push  [%,d] segments.", segmentsToPush.size()), e);
+      Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public void write(Writable w) throws IOException {
+    if (w == null) {
```

Comment:
Will this ever happen? If it will, when? Should we document it and throw a WARN?